src/modules/main/PlayerView.tsx

```diff
@@ -1,23 +1,159 @@
+
+type SubmittedAnswer = {
+  expectedId: string;
+  actualId?: string;
+};
 
 const PlayerView = (): JSX.Element => {
-  const { permission } = useLocalContext();
   const { data: appContext } = hooks.useAppContext();
-  const members = appContext?.members;
+  const { t } = useAppTranslation();
+  const { data: appData } = hooks.useAppData<{ answers: SubmittedAnswer[] }>();
+  const { data: appSettings, isLoading } = hooks.useAppSettings<Settings>({
+    name: SettingsKeys.SettingsData,
+  });
+  const { mutate: saveAppData } = mutations.usePostAppData();
+  const [answeredLabels, setAnsweredLabels] = useState<AnsweredLabel[]>([]);
+  // labels will be null only before setting the state as we cannot render all labels within container if not settled yet
+  const [labels, setLabels] = useState<null | Label[]>(null);
+
+  const { data: image } = hooks.useAppSettings({
+    name: SettingsKeys.File,
+  });
+  const [isSubmitted, setIsSubmitted] = useState(false);
+
+  const answersAppData = appData?.filter(
+    ({ type }) => type === ANSWER_SUBMISSION_TYPE,
+  );
+  const settingLabels = appSettings?.[0].data.labels;
+  const lastVersion = answersAppData?.[answersAppData.length - 1];
+  const answers = lastVersion?.data.answers;
+
+  const retry = (): void => {
+    if (settingLabels) {
+      const answered = settingLabels?.map((label) => ({
+        expected: label,
+        actual: null,
+      }));
+
+      setAnsweredLabels(answered);
+      setLabels(settingLabels);
+    }
+  };
+
+  const save = (): void => {
+    const submittedAnswers = answeredLabels.map(({ expected, actual }) => ({
+      expectedId: expected.id,
+      actualId: actual?.id,
+    }));
+    saveAppData({
+      data: { answers: submittedAnswers },
+      type: ANSWER_SUBMISSION_TYPE,
+    });
+  };
+
+  useEffect(() => {
+    if (settingLabels) {
+      if (answers) {
+        const answered = answers?.map(
+          ({ expectedId, actualId }: SubmittedAnswer) => ({
+            expected: settingLabels.find(
+              ({ id }) => id === expectedId,
+            ) as Label,
+            actual: settingLabels.find(({ id }) => id === actualId) || null,
+          }),
+        );
+
+        setAnsweredLabels(answered);
+        setLabels(
+          settingLabels.filter(
+            ({ id }) =>
+              !answers.find(({ actualId }: SubmittedAnswer) => actualId === id),
+          ),
+        );
+      } else {
+        const answered = settingLabels.map((label) => ({
+          expected: label,
+          actual: null,
+        }));
+
+        setAnsweredLabels(answered);
+        setLabels(settingLabels);
+      }
+    }
+  }, [answers, settingLabels]);
+
+  useEffect(() => {
+    const answered = answers?.map(
+      ({ expectedId, actualId }: SubmittedAnswer) => ({
+        expected: settingLabels?.find(({ id }) => id === expectedId) as Label,
+        actual: settingLabels?.find(({ id }) => id === actualId) || null,
+      }),
+    );
```

Comment:
You also have this in the other useEffect, do you still need it here? Maybe you could extract the map function to reuse it if it is still needed?

Reply:
I wonder if it could be in a context indeed. Maybe be careful with the re-rendering as well.